tests/5.0/target_data/test_target_data_use_device_addr.c

```diff
@@ -0,0 +1,51 @@
+//===-- test_target_data_use_device_addr.c - test of use_device_addr on target data ----===//
+// 
+// OpenMP API Version 5.0 Nov 2018
+// 
+// This file is a test for the use_device_addr when used with the map
+// clause with target data directive. This test uses a scalar and an array of size N 
+// which values are modified on the  device and tested in the host. 
+// List items that appear in a use_device_addr clause have the address 
+// of the corresponding object in the device data environment inside the construct. 
+// This test also tests that address conversions of use_device_addr clauses will 
+// occur as if performed after all variables are mapped according to those map clauses.
+//===----------------------------------------------------------------------===//
+
+#include <omp.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include "ompvv.h"
+
+
+int main() {
+  int errors = 0;
+  int device_data = 14, host_data=0;
+
+  OMPVV_TEST_OFFLOADING;
+
+#pragma omp target data map(to: device_data) use_device_addr(device_data)
+    {
+        int *dev_ptr;
+        dev_ptr = &device_data;
+#pragma omp target map(to:device_data) map(tofrom: errors) map(from: host_data)
+      {
```

Comment:
I think you need a `is_device_ptr(dev_ptr)` here.

Additionally, I fear the compiler misunderstands the `map(to:device_data)` by trying to map the device-address `device_data` instead of the host-address one. At least when trying to compile it with GCC it fails for `&device_data != dev_ptr` and `device_addr` has an odd address.

I think you need something like:

```
int *dev_ptr;
// ↓ map data – could be either 'target enter/exit data' as here or another 'target data' encompassing everything
#pragma omp target enter data map(to:device_data)
#pragma omp target data use_device_addr(device_data)
{
         dev_ptr = &device_data;
}
// ↓ this target is outside the 'use_device_addr' (!); additionally 'is_device_ptr' has been added.
#pragma omp target is_device_ptr(dev_ptr) map(to:device_data) map(tofrom: errors) map(from: host_data)
      {
           if(&device_data != dev_ptr)
…

// At the very end: be kind and clean up  ('release:' would also do)
#pragma omp target exit data map(delete: device_data)
```

Reply:
It will be interesting to find out which address it sees for 'device_addr'